File: Sources/CurioCopy/Format.swift
// Hand-authored (not generated) — the Swift mirror of src/format.ts's canonical spec. See that
// file's header comment for the full spec; this file must implement it identically. Both
// implementations are tested against the SAME fixtures (src/fixtures.json / Tests/CurioCopyTests/
// FormatTests.swift) so a divergence shows up as a failing test in whichever repo hits it first.

import Foundation

extension CurioCopy {
    /// "£245.00" / "-£12.34" / "£1,234.56" — en-GB, always 2dp, "," thousands separator, "£"
    /// prefixed with no space, minus sign before the £ on negative amounts.
    public static func formatGBP(_ amount: Double) -> String {
        guard amount.isFinite else { return "£0.00" }
        let negative = amount < 0
        let abs = Swift.abs(amount)
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "en_GB")
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        formatter.groupingSeparator = ","
        formatter.decimalSeparator = "."
        let formatted = formatter.string(from: NSNumber(value: abs)) ?? String(format: "%.2f", abs)
        return "\(negative ? "-" : "")£\(formatted)"
    }

    private static let shortMonths = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

    /// "9 Aug" (style .short, default) or "9 Aug 2026" (style .full). Empty string on unparseable
    /// input — a formatter must never crash a render for bad data.
    public enum DateStyle { case short, full }

    public static func formatDate(_ isoDateString: String, style: DateStyle = .short) -> String {
        // UTC, deliberately — a plain "YYYY-MM-DD" input is a calendar date, not an instant; local
        // getters would shift the day depending on the runtime's timezone. Matches format.ts.
        var utcCalendar = Calendar(identifier: .gregorian)
        utcCalendar.timeZone = TimeZone(identifier: "UTC")!
        let isoFormatter = ISO8601DateFormatter()
        isoFormatter.formatOptions = [.withFullDate]
        guard let date = isoFormatter.date(from: isoDateString) ?? fallbackParse(isoDateString) else { return "" }
        let comps = utcCalendar.dateComponents([.day, .month, .year], from: date)
        guard let day = comps.day, let month = comps.month, let year = comps.year else { return "" }
        let monthName = shortMonths[month - 1]
        return style == .full ? "\(day) \(monthName) \(year)" : "\(day) \(monthName)"
    }

    /// ISO8601DateFormatter with .withFullDate only accepts exactly "YYYY-MM-DD"; fall back to a
    /// plain DateFormatter for inputs with a time component, same tolerance `new Date(...)` has in
    /// the TS implementation.
    private static func fallbackParse(_ s: String) -> Date? {
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
        df.timeZone = TimeZone(identifier: "UTC")
        if let d = df.date(from: s) { return d }
        df.dateFormat = "yyyy-MM-dd"
        df.timeZone = TimeZone(identifier: "UTC")
        return df.date(from: s)
    }
}

File: src/format.ts
// Canonical en-GB currency + date formatting spec (WORK-BACKLOG.md Packet CC-1 slice 1).
//
// pokemon-tool had this same £-formatter hand-defined 17 times (as a local `money()` helper) —
// this file is the one implementation. There is no way to share compiled JS with Swift, so the
// SPEC below is implemented twice — once here, once in Sources/CurioCopy/Format.swift — but both
// implementations are exact, deliberate mirrors of the same rules, and `fixtures.json` in this
// directory is the shared input/output truth table both platforms' test suites assert against.
// If you change a rule, update BOTH implementations and the fixtures in the same change.
//
// ── SPEC ──────────────────────────────────────────────────────────────────────────────────────
// formatGBP(amount):
//   - Locale: en-GB. Currency: GBP. Symbol: "£", prefixed (no space).
//   - Always exactly 2 decimal places.
//   - Thousands grouped with "," (e.g. 1,234.56).
//   - Negative amounts: "-£12.34" (minus sign before the £, not after).
// formatDate(isoDateString, style):
//   - style "short" (default): "9 Aug" — day (no leading zero) + space + short month name, en-GB.
//   - style "full": "9 Aug 2026" — adds the year.
//   - Invalid/unparseable input returns "" rather than throwing (a formatter must never crash a
//     render for bad data — surface the missing date some other way, not by throwing here).

export function formatGBP(amount: number): string {
  if (!Number.isFinite(amount)) return "£0.00";
  const negative = amount < 0;
  const abs = Math.abs(amount);
  const formatted = abs.toLocaleString("en-GB", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  return `${negative ? "-" : ""}£${formatted}`;
}

const SHORT_MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

export function formatDate(isoDateString: string, style: "short" | "full" = "short"): string {
  const d = new Date(isoDateString);
  if (Number.isNaN(d.getTime())) return "";
  // UTC getters, deliberately — a plain "YYYY-MM-DD" input is a calendar date, not an instant, so
  // reading it back with LOCAL getters would shift the day depending on the runtime's timezone
  // (e.g. "2026-08-09" parses as UTC midnight; a UTC-8 runtime's local getDate() would read back
  // Aug 8). UTC getters make this deterministic regardless of where it runs.
  const day = d.getUTCDate();
  const month = SHORT_MONTHS[d.getUTCMonth()];
  return style === "full" ? `${day} ${month} ${d.getUTCFullYear()}` : `${day} ${month}`;
}
